Add support for GetLocaleInfoW

This is used by ee-as.exe included as part of ProDG's PS2 toolchain (found [here](https://archive.org/download/SNSystemsProDGPs2/ProDGPs2usrLocalSceFiles.zip)).

Current example output:
```
wibo /usr/local/prodg/ee/gcc/bin/ee-as.exe -c -EL -mcpu=5900 -32 -g2 -non_shared -G0 -I include -o build/asm/crt0.s.o asm/crt0.s
Unhandled function GetLocaleInfoW (KERNEL32.dll)
make: *** [Makefile:70: build/asm/crt0.s.o] Error 1
```

My Dark Cloud decompilation project makes use of ee-as as per https://github.com/Adubbz/DCDecomp/blob/master/Makefile#L27 (currently using wine rather than wibo).

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,10 +28,11 @@ add_executable(wibo
     dll/vcruntime.cpp
     dll/version.cpp
     files.cpp
-    processes.cpp
     handles.cpp
     loader.cpp
     main.cpp
+    processes.cpp
+    strutil.cpp
 )
 target_link_libraries(wibo PRIVATE std::filesystem)
 install(TARGETS wibo DESTINATION bin)

--- a/common.h
+++ b/common.h
@@ -6,6 +6,7 @@
 #include <string>
 #include <sys/stat.h>
 #include <unistd.h>
+#include <vector>
 
 // On Windows, the incoming stack is aligned to a 4 byte boundary.
 // force_align_arg_pointer will realign the stack to match GCC's 16 byte alignment.
@@ -74,6 +75,7 @@ namespace wibo {
 	extern int argc;
 	extern char *executableName;
 	extern char *commandLine;
+	extern std::vector<uint16_t> commandLineW;
 	extern bool debugEnabled;
 	extern unsigned int debugIndent;
 

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -1791,7 +1791,7 @@ namespace kernel32 {
 		DEBUG_LOG("GetLocaleInfoW %d %d\n", Locale, LCType);
 		std::string info = str_for_LCType(LCType);
 		auto ret = stringToWideString(info.c_str());
-		size_t len = info.size() + 1;
+		size_t len = ret.size();
 
 		if (!cchData) {
 			return len;
@@ -1860,10 +1860,11 @@ namespace kernel32 {
 		}
 		auto wideValue = stringToWideString(value);
 		const auto len = wstrlen(wideValue.data());
-		if (nSize < len + 1) {
-			return len + 1;
+		const auto wvSize = wideValue.size();
+		if (nSize < wvSize) {
+			return wvSize;
 		}
-		wstrncpy(lpBuffer, wideValue.data(), len + 1);
+		wstrncpy(lpBuffer, wideValue.data(), wvSize);
 		return len;
 	}
 

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -1800,8 +1800,7 @@ namespace kernel32 {
 		return 1;
 	}
 
-	int WIN_FUNC GetLocaleInfoA(unsigned int Locale, int LCType, char *lpLCData, int cchData) {
-		DEBUG_LOG("GetLocaleInfoA %d %d\n", Locale, LCType);
+	std::string str_for_LCType(int LCType) {
 		std::string ret;
 		// https://www.pinvoke.net/default.aspx/Enums/LCType.html
 		if (LCType == 4100) { // LOCALE_IDEFAULTANSICODEPAGE
@@ -1814,6 +1813,12 @@ namespace kernel32 {
 		if (LCType == 4098) { // LOCALE_SENGCOUNTRY
 			ret = "Country";
 		}
+		return ret;
+	}
+
+	int WIN_FUNC GetLocaleInfoA(unsigned int Locale, int LCType, char *lpLCData, int cchData) {
+		DEBUG_LOG("GetLocaleInfoA %d %d\n", Locale, LCType);
+		std::string ret = str_for_LCType(LCType);
 
 		if (!cchData) {
 			return ret.size() + 1;
@@ -1823,8 +1828,17 @@ namespace kernel32 {
 		}
 	}
 
-	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, char *lpLCData, int cchData) {
-		return GetLocaleInfoA(Locale, LCType, lpLCData, cchData); // lol
+	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, wchar_t *lpLCData, int cchData) {
+		DEBUG_LOG("GetLocaleInfoW %d %d\n", Locale, LCType);
+		std::string info = str_for_LCType(LCType);
+		std::wstring ret(ret.begin(), ret.end());
+
+		if (!cchData) {
+			return ret.size() + 1;
+		} else {
+			memcpy(lpLCData, ret.c_str(), ret.size() + sizeof(wchar_t));
+			return 1;
+		}
 	}
 
 	int WIN_FUNC GetUserDefaultLCID() {

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -1823,6 +1823,10 @@ namespace kernel32 {
 		}
 	}
 
+	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, char *lpLCData, int cchData) {
+		return GetLocaleInfoA(Locale, LCType, lpLCData, cchData); // lol
+	}
+
 	int WIN_FUNC GetUserDefaultLCID() {
 		DEBUG_LOG("GetUserDefaultLCID\n");
 		return 1;
@@ -2007,6 +2011,7 @@ static void *resolveByName(const char *name) {
 	if (strcmp(name, "LCMapStringW") == 0) return (void *) kernel32::LCMapStringW;
 	if (strcmp(name, "LCMapStringA") == 0) return (void *) kernel32::LCMapStringA;
 	if (strcmp(name, "GetLocaleInfoA") == 0) return (void *) kernel32::GetLocaleInfoA;
+	if (strcmp(name, "GetLocaleInfoW") == 0) return (void *) kernel32::GetLocaleInfoW;
 	if (strcmp(name, "GetUserDefaultLCID") == 0) return (void *) kernel32::GetUserDefaultLCID;
 	if (strcmp(name, "IsDBCSLeadByte") == 0) return (void *) kernel32::IsDBCSLeadByte;
 

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -8,6 +8,7 @@
 #include <filesystem>
 #include <fnmatch.h>
 #include <string>
+#include "strutil.h"
 #include <malloc.h>
 #include <stdarg.h>
 #include <system_error>
@@ -39,24 +40,6 @@ typedef struct _EXCEPTION_POINTERS {
 typedef LONG (*PVECTORED_EXCEPTION_HANDLER)(PEXCEPTION_POINTERS ExceptionInfo);
 
 namespace kernel32 {
-	static size_t wstrlen(const uint16_t *str) {
-		size_t len = 0;
-		while (str[len] != 0)
-			++len;
-		return len;
-	}
-
-	static size_t wstrncpy(uint16_t *dst, const uint16_t *src, size_t n) {
-		size_t i = 0;
-		while (i < n && src[i] != 0) {
-			dst[i] = src[i];
-			++i;
-		}
-		if (i < n)
-			dst[i] = 0;
-		return i;
-	}
-
 	static void *doAlloc(unsigned int dwBytes, bool zero) {
 		if (dwBytes == 0)
 			dwBytes = 1;
@@ -77,29 +60,6 @@ namespace kernel32 {
 			memset((char*)ret + oldSize, 0, newSize - oldSize);
 		}
 		return ret;
-	}
-
-	static std::string wideStringToString(const uint16_t *src, int len = -1) {
-		if (len < 0) {
-			len = src ? wstrlen(src) : 0;
-		}
-		std::string res(len, '\0');
-		for (int i = 0; i < len; i++) {
-			res[i] = src[i] & 0xFF;
-		}
-		return res;
-	}
-
-	static std::vector<uint16_t> stringToWideString(const char *src) {
-		int len = strlen(src);
-		std::vector<uint16_t> res(len + 1);
-
-		for (size_t i = 0; i < res.size(); i++) {
-			res[i] = src[i] & 0xFF;
-		}
-		res[len] = 0; // NUL terminate
-
-		return res;
 	}
 
 	static int doCompareString(const std::string &a, const std::string &b, unsigned int dwCmpFlags) {
@@ -477,7 +437,7 @@ namespace kernel32 {
 
 	LPWSTR WIN_FUNC GetCommandLineW() {
 		DEBUG_LOG("GetCommandLineW -> ");
-		return stringToWideString(GetCommandLineA()).data();
+		return wibo::commandLineW.data();
 	}
 
 	char *WIN_FUNC GetEnvironmentStrings() {
@@ -624,14 +584,14 @@ namespace kernel32 {
 		const auto lpFileNameA = wideStringToString(lpFileName);
 		std::filesystem::path absPath = std::filesystem::absolute(files::pathFromWindows(lpFileNameA.c_str()));
 		std::string absStr = files::pathToWindows(absPath);
-		const auto absStrW = stringToWideString(absStr.c_str()).data();
+		const auto absStrW = stringToWideString(absStr.c_str());
 		DEBUG_LOG("-> %s\n", absStr.c_str());
 
-		const auto len = wstrlen(absStrW);
+		const auto len = wstrlen(absStrW.data());
 		if (nBufferLength < len + 1) {
 			return len + 1;
 		}
-		wstrncpy(lpBuffer, absStrW, len + 1);
+		wstrncpy(lpBuffer, absStrW.data(), len + 1);
 		assert(!lpFilePart);
 		return len;
 	}
@@ -1821,7 +1781,7 @@ namespace kernel32 {
 		if (!cchData) {
 			return len;
 		} else {
-			assert(len <= cchData);
+			assert(len <= (size_t) cchData);
 			memcpy(lpLCData, ret.c_str(), len);
 			return 1;
 		}
@@ -1830,14 +1790,14 @@ namespace kernel32 {
 	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, LPWSTR lpLCData, int cchData) {
 		DEBUG_LOG("GetLocaleInfoW %d %d\n", Locale, LCType);
 		std::string info = str_for_LCType(LCType);
-		LPWSTR ret = stringToWideString(info.c_str()).data();
+		auto ret = stringToWideString(info.c_str());
 		size_t len = info.size() + 1;
 
 		if (!cchData) {
 			return len;
 		} else {
-			assert(len <= cchData);
-			memcpy(lpLCData, ret, len * sizeof(*ret));
+			assert(len <= (size_t) cchData);
+			memcpy(lpLCData, ret.data(), len * sizeof(*ret.data()));
 			return 1;
 		}
 	}
@@ -1898,12 +1858,12 @@ namespace kernel32 {
 		if (!value) {
 			return 0;
 		}
-		uint16_t *wideValue = stringToWideString(value).data();
-		const auto len = wstrlen(wideValue);
+		auto wideValue = stringToWideString(value);
+		const auto len = wstrlen(wideValue.data());
 		if (nSize < len + 1) {
 			return len + 1;
 		}
-		wstrncpy(lpBuffer, wideValue, len + 1);
+		wstrncpy(lpBuffer, wideValue.data(), len + 1);
 		return len;
 	}
 

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -1859,13 +1859,12 @@ namespace kernel32 {
 			return 0;
 		}
 		auto wideValue = stringToWideString(value);
-		const auto len = wstrlen(wideValue.data());
-		const auto wvSize = wideValue.size();
-		if (nSize < wvSize) {
-			return wvSize;
+		const auto len = wideValue.size();
+		if (nSize < len) {
+			return len;
 		}
-		wstrncpy(lpBuffer, wideValue.data(), wvSize);
-		return len;
+		wstrncpy(lpBuffer, wideValue.data(), len);
+		return len - 1;
 	}
 
 	unsigned int WIN_FUNC QueryPerformanceCounter(unsigned long int *lpPerformanceCount) {

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -15,6 +15,7 @@
 #include <sys/stat.h>
 #include <sys/wait.h>
 #include <spawn.h>
+#include <vector>
 
 typedef union _RTL_RUN_ONCE {
 	PVOID Ptr;
@@ -89,12 +90,11 @@ namespace kernel32 {
 		return res;
 	}
 
-	static uint16_t *stringToWideString(const char *src) {
-		uint16_t *res = nullptr;
-
+	static std::vector<uint16_t> stringToWideString(const char *src) {
 		int len = strlen(src);
-		res = (uint16_t *)malloc((len + 1) * 2);
-		for (int i = 0; i < len; i++) {
+		std::vector<uint16_t> res(len + 1);
+
+		for (size_t i = 0; i < res.size(); i++) {
 			res[i] = src[i] & 0xFF;
 		}
 		res[len] = 0; // NUL terminate
@@ -477,7 +477,7 @@ namespace kernel32 {
 
 	LPWSTR WIN_FUNC GetCommandLineW() {
 		DEBUG_LOG("GetCommandLineW -> ");
-		return stringToWideString(GetCommandLineA());
+		return stringToWideString(GetCommandLineA()).data();
 	}
 
 	char *WIN_FUNC GetEnvironmentStrings() {
@@ -624,17 +624,15 @@ namespace kernel32 {
 		const auto lpFileNameA = wideStringToString(lpFileName);
 		std::filesystem::path absPath = std::filesystem::absolute(files::pathFromWindows(lpFileNameA.c_str()));
 		std::string absStr = files::pathToWindows(absPath);
-		const auto absStrW = stringToWideString(absStr.c_str());
+		const auto absStrW = stringToWideString(absStr.c_str()).data();
 		DEBUG_LOG("-> %s\n", absStr.c_str());
 
 		const auto len = wstrlen(absStrW);
 		if (nBufferLength < len + 1) {
-			free(absStrW);
 			return len + 1;
 		}
 		wstrncpy(lpBuffer, absStrW, len + 1);
 		assert(!lpFilePart);
-		free(absStrW);
 		return len;
 	}
 
@@ -1812,7 +1810,7 @@ namespace kernel32 {
 		if (LCType == 4098) { // LOCALE_SENGCOUNTRY
 			return "Country";
 		}
-		assert(false); // something better?
+		assert(false);
 	}
 
 	int WIN_FUNC GetLocaleInfoA(unsigned int Locale, int LCType, LPSTR lpLCData, int cchData) {
@@ -1832,14 +1830,14 @@ namespace kernel32 {
 	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, LPWSTR lpLCData, int cchData) {
 		DEBUG_LOG("GetLocaleInfoW %d %d\n", Locale, LCType);
 		std::string info = str_for_LCType(LCType);
-		LPWSTR ret = stringToWideString(info.c_str());
-		size_t len = (info.size() + 1) * sizeof(uint16_t);
+		LPWSTR ret = stringToWideString(info.c_str()).data();
+		size_t len = info.size() + 1;
 
 		if (!cchData) {
 			return len;
 		} else {
 			assert(len <= cchData);
-			memcpy(lpLCData, ret, len);
+			memcpy(lpLCData, ret, len * sizeof(*ret));
 			return 1;
 		}
 	}
@@ -1900,14 +1898,12 @@ namespace kernel32 {
 		if (!value) {
 			return 0;
 		}
-		uint16_t *wideValue = stringToWideString(value);
+		uint16_t *wideValue = stringToWideString(value).data();
 		const auto len = wstrlen(wideValue);
 		if (nSize < len + 1) {
-			free(wideValue);
 			return len + 1;
 		}
 		wstrncpy(lpBuffer, wideValue, len + 1);
-		free(wideValue);
 		return len;
 	}
 

--- a/dll/kernel32.cpp
+++ b/dll/kernel32.cpp
@@ -1801,42 +1801,45 @@ namespace kernel32 {
 	}
 
 	std::string str_for_LCType(int LCType) {
-		std::string ret;
 		// https://www.pinvoke.net/default.aspx/Enums/LCType.html
 		if (LCType == 4100) { // LOCALE_IDEFAULTANSICODEPAGE
 			// Latin1; ref GetACP
-			ret = "28591";
+			return "28591";
 		}
 		if (LCType == 4097) { // LOCALE_SENGLANGUAGE
-			ret = "Lang";
+			return "Lang";
 		}
 		if (LCType == 4098) { // LOCALE_SENGCOUNTRY
-			ret = "Country";
+			return "Country";
 		}
-		return ret;
+		assert(false); // something better?
 	}
 
-	int WIN_FUNC GetLocaleInfoA(unsigned int Locale, int LCType, char *lpLCData, int cchData) {
+	int WIN_FUNC GetLocaleInfoA(unsigned int Locale, int LCType, LPSTR lpLCData, int cchData) {
 		DEBUG_LOG("GetLocaleInfoA %d %d\n", Locale, LCType);
 		std::string ret = str_for_LCType(LCType);
+		size_t len = ret.size() + 1;
 
 		if (!cchData) {
-			return ret.size() + 1;
+			return len;
 		} else {
-			memcpy(lpLCData, ret.c_str(), ret.size() + 1);
+			assert(len <= cchData);
+			memcpy(lpLCData, ret.c_str(), len);
 			return 1;
 		}
 	}
 
-	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, wchar_t *lpLCData, int cchData) {
+	int WIN_FUNC GetLocaleInfoW(unsigned int Locale, int LCType, LPWSTR lpLCData, int cchData) {
 		DEBUG_LOG("GetLocaleInfoW %d %d\n", Locale, LCType);
 		std::string info = str_for_LCType(LCType);
-		std::wstring ret(ret.begin(), ret.end());
+		LPWSTR ret = stringToWideString(info.c_str());
+		size_t len = (info.size() + 1) * sizeof(uint16_t);
 
 		if (!cchData) {
-			return ret.size() + 1;
+			return len;
 		} else {
-			memcpy(lpLCData, ret.c_str(), ret.size() + sizeof(wchar_t));
+			assert(len <= cchData);
+			memcpy(lpLCData, ret, len);
 			return 1;
 		}
 	}

--- a/main.cpp
+++ b/main.cpp
@@ -4,17 +4,20 @@
 #include <filesystem>
 #include <errno.h>
 #include <memory>
+#include "strutil.h"
 #include <sys/mman.h>
 #include <sys/syscall.h>
 #include <stdarg.h>
 #include <iostream>
 #include <fstream>
+#include <vector>
 
 uint32_t wibo::lastError = 0;
 char** wibo::argv;
 int wibo::argc;
 char *wibo::executableName;
 char *wibo::commandLine;
+std::vector<uint16_t> wibo::commandLineW;
 wibo::Executable *wibo::mainModule = 0;
 bool wibo::debugEnabled = false;
 unsigned int wibo::debugIndent = 0;
@@ -274,6 +277,7 @@ int main(int argc, char **argv) {
 	cmdLine += '\0';
 
 	wibo::commandLine = cmdLine.data();
+	wibo::commandLineW = stringToWideString(wibo::commandLine);
 	DEBUG_LOG("Command line: %s\n", wibo::commandLine);
 
 	wibo::executableName = argv[0];

--- a/strutil.cpp
+++ b/strutil.cpp
@@ -1,0 +1,44 @@
+#include "common.h"
+#include "strings.h"
+#include <vector>
+
+size_t wstrlen(const uint16_t *str) {
+	size_t len = 0;
+	while (str[len] != 0)
+		++len;
+	return len;
+}
+
+size_t wstrncpy(uint16_t *dst, const uint16_t *src, size_t n) {
+	size_t i = 0;
+	while (i < n && src[i] != 0) {
+		dst[i] = src[i];
+		++i;
+	}
+	if (i < n)
+		dst[i] = 0;
+	return i;
+}
+
+std::string wideStringToString(const uint16_t *src, int len = -1) {
+	if (len < 0) {
+		len = src ? wstrlen(src) : 0;
+	}
+	std::string res(len, '\0');
+	for (int i = 0; i < len; i++) {
+		res[i] = src[i] & 0xFF;
+	}
+	return res;
+}
+
+std::vector<uint16_t> stringToWideString(const char *src) {
+	int len = strlen(src);
+	std::vector<uint16_t> res(len + 1);
+
+	for (size_t i = 0; i < res.size(); i++) {
+		res[i] = src[i] & 0xFF;
+	}
+	res[len] = 0; // NUL terminate
+
+	return res;
+}

--- a/strutil.h
+++ b/strutil.h
@@ -1,0 +1,7 @@
+#include <string>
+#include <vector>
+
+size_t wstrlen(const uint16_t *str);
+size_t wstrncpy(uint16_t *dst, const uint16_t *src, size_t n);
+std::string wideStringToString(const uint16_t *src, int len = -1);
+std::vector<uint16_t> stringToWideString(const char *src);

